File: JULY-2025/28-07-2025/p2044.kt
class Solution {
    fun countMaxOrSubsets(nums: IntArray): Int {
        var max = 0
        for (num in nums) {
            max = max or num
        }
        var res = 0
        val n = nums.size
        for (i in 0..n - 1) {
            res += dfs(i, 0, max, nums)
        }
        return res
    }

    private fun dfs(pos: Int, cur: Int, max: Int, nums: IntArray): Int {
        val n = nums.size
        var res = 0
        if (cur or nums[pos] == max) {
            res += 1 shl n - pos - 1
        } else {
            for (i in pos + 1..n - 1) {
                res += dfs(i, cur or nums[pos], max, nums)
            }
        }
        return res
    }
}

File: JULY-2025/28-07-2025/p2044.cs
public class Solution {
    // { OR result : freq }
    SortedDictionary<int, int> descDict = new(Comparer<int>.Create((x, y) => y.CompareTo(x)));

    public int CountMaxOrSubsets(int[] nums) {
        // generate all subsets
        // init prevCal = 0 because x | 0 = x
        dfs(0, nums, 0);
        return descDict.Values.First();
    }

    public int dfs(int index, int[] nums, int prevCal){
        // HERE IS THE CATCH
        // max OR will be the OR of all elements in subarr
        // so we will add the local result when it reach the end of the subarray
        if(index >= nums.Length){
            if(descDict.ContainsKey(prevCal)){
                descDict[prevCal] += 1;
            }else
            {
                descDict[prevCal] = 1;
            }
            return prevCal;
        }

        // decision 1, include nums[i]
        int newCal1 = dfs(index + 1, nums, prevCal | nums[index]);

        // decision 2, not to include nums[i];
        int newCal2 = dfs(index + 1, nums, prevCal);
        
        // finding max OR bitwise
        return Math.Max(newCal1, newCal2);
    }
}